File: programmers/이분탐색/입국심사.cpp
#include <string>
#include <vector>

#define INF 1000000000000000000
using namespace std;
long long startT, endT = INF, totalTime = INF;

long long solution(int n, vector<int> times) {
    long long answer = INF;
    
    while(startT <= endT){
        long long mid = (startT + endT) / 2;
        
        long long cnt = 0;
        for(int i = 0; i < times.size(); i++){
            cnt += mid/times[i];
        }
        
        if(cnt >= n){
            answer = min(answer, mid);
            endT = mid - 1;
        }else{
            startT = mid + 1;
        }
        
    }
    
    return answer;
}

File: python/이분탐색/입국심사.py
#https://school.programmers.co.kr/learn/courses/30/lessons/43238?language=python3
def solution(n, times):
    max_time = max(times) * n
    answer = max_time
    s = 1
    e = max_time
    
    while s <= e:
        mid = (s+e)//2
        
        cnt = 0
        for tim in times:
            cnt += mid//tim
        
        if cnt >= n:
            answer = min(answer, mid)
            e = mid - 1
        else:
            s = mid + 1
    
    return answer